crates/bootstrap_srv/src/http.rs

```diff
@@ -0,0 +1,254 @@
+type HRecv = async_channel::Receiver<(HttpRequest, HttpRespondCb)>;
+
+#[derive(Clone)]
+pub struct HttpReceiver(HRecv);
+
+impl HttpReceiver {
+    pub fn recv(&self) -> Option<(HttpRequest, HttpRespondCb)> {
+        match self.0.recv_blocking() {
+            Ok(r) => Some(r),
+            Err(_) => None,
+        }
+    }
+}
+
+pub struct ServerConfig {
+    pub addr: std::net::SocketAddr,
+    pub worker_thread_count: usize,
+}
+
+pub struct Server {
+    t_join: Option<std::thread::JoinHandle<()>>,
+    addr: std::net::SocketAddr,
+    receiver: HttpReceiver,
+    h_send: HSend,
+    shutdown: Option<tokio::sync::oneshot::Sender<()>>,
+}
+
+impl Drop for Server {
+    fn drop(&mut self) {
+        self.h_send.close();
+        if let Some(shutdown) = self.shutdown.take() {
+            let _ = shutdown.send(());
+        }
+        if let Some(t_join) = self.t_join.take() {
+            let _ = t_join.join();
+        }
+    }
+}
+
+impl Server {
+    pub fn new(config: ServerConfig) -> Result<Self> {
+        let (s_ready, r_ready) = tokio::sync::oneshot::channel();
+        let t_join = std::thread::spawn(move || tokio_thread(config, s_ready));
+        match r_ready.blocking_recv() {
+            Ok(Ok(Ready {
+                h_send,
+                addr,
+                receiver,
+                shutdown,
+            })) => Ok(Self {
+                t_join: Some(t_join),
+                addr,
+                receiver,
+                h_send,
+                shutdown: Some(shutdown),
+            }),
+            Ok(Err(err)) => Err(err),
+            Err(_) => Err(std::io::Error::other("failed to bind server")),
+        }
+    }
+
+    pub fn server_addr(&self) -> std::net::SocketAddr {
+        self.addr
+    }
+
+    pub fn receiver(&self) -> &HttpReceiver {
+        &self.receiver
+    }
+}
+
+struct Ready {
+    h_send: HSend,
+    addr: std::net::SocketAddr,
+    receiver: HttpReceiver,
+    shutdown: tokio::sync::oneshot::Sender<()>,
+}
+
+fn tokio_thread(
+    config: ServerConfig,
+    ready: tokio::sync::oneshot::Sender<std::io::Result<Ready>>,
+) {
+    tokio::runtime::Builder::new_multi_thread()
+        .enable_all()
+        .build()
+        .unwrap()
+        .block_on(async move {
+            let (h_send, h_recv) =
+                async_channel::bounded(config.worker_thread_count);
+
+            let app: Router = Router::new()
+                .route("/health", routing::get(handle_health_get))
+                .route("/bootstrap/:space", routing::get(handle_boot_get))
+                .route(
+                    "/bootstrap/:space/:agent",
+                    routing::put(handle_boot_put),
+                )
+                .layer(extract::DefaultBodyLimit::max(1024))
+                .with_state(h_send.clone());
+
+            let receiver = HttpReceiver(h_recv);
+
+            let listener =
+                match tokio::net::TcpListener::bind(config.addr).await {
```

Comment:
It was less obvious before and I missed asking about it, but ipv4/v6. Does supporting both matter here?

Reply:
Yes, this is another area where tiny_http was immature. It let you specify multiple bindings but only reported a single one, so was not super useful... I'll update the api to be more reasonable now that we can.

Reply:
Done: https://github.com/holochain/kitsune2/pull/52/commits/a08fdf59d9acac72434cea726d800d6a99152db7